group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -983,27 +1052,125 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
         return new CoordinatorResult<>(records, response);
     }
 
+    private void replaceStaticMemberInConsumerGroup(
+        String groupId,
+        List<Record> records,
+        ConsumerGroup group,
+        ConsumerGroupMember member,
+        ConsumerGroupMember existingStaticMember,
+        ConsumerGroupMember updatedMember
+    ) {
+        // Write tombstones for the departed static member. Follow the same order as the one followed when fencing
+        // a member
+        records.add(newCurrentAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newTargetAssignmentTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        records.add(newMemberSubscriptionTombstoneRecord(group.groupId(), existingStaticMember.memberId()));
+        // Cancel all the timers of the departed static member.
+        cancelConsumerGroupSessionTimeout(group.groupId(), existingStaticMember.memberId());
+        cancelConsumerGroupRevocationTimeout(group.groupId(), existingStaticMember.memberId());
+        // Write a record corresponding to the new member
+        records.add(newMemberSubscriptionRecord(groupId, updatedMember));
+        TargetAssignmentBuilder.TargetAssignmentResult assignmentResult = computeTargetAssignment(group, group.groupEpoch(), member, updatedMember);
+        records.addAll(assignmentResult.records());
+        records.add(newCurrentAssignmentRecord(groupId, updatedMember));
```

Comment:
I am not sold on this. Is it too difficult to reuse the main logic? There are a few issues with this approach. For instance, the member's assignment is not reconciled like we do in the main logic. Another one is that the subscription metadata must be updated as well if the subscriptions have changed.

Reply:
The main reason for extracting this out was that while using the main logic, there always always a group epoch bump even when a new static member replaces an older one. When I debugged it further, it seems to be because of this logic [here](https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L885-L889)
. More specifically, the issue at hand is that `subscriptionMetadata` has partition racks info while the currently stored metadata doesn't have it. This is with regards to the test `testStaticMemberGetsBackAssignmentUponRejoin`. I  wasn't totally sure if this is an issue with the test itself but since this led to a group epoch bump, I thought we shouldn't do it. 

Actually when I think about it now, maybe it makes sense to have a group epoch bump in this case as well. While it might go against no rebalance during static member rejoin but the reason for rejoin is a change in subscription metadata and not a static member re-join. The latter seemed harder to replicate via tests though because it always bumped up the group epoch due to the above mentioned issue. Please let me know your thoughts.

Reply:
Hum... I am not sure to fully follow. The subscription metadata should not be different if the subscriptions and the metadata image have not changed. Does the new static member has the same subs as the previous one in your case?

Reply:
I got around the issue by explicitly setting the rack info in the subscription metadata like [here](https://github.com/apache/kafka/pull/14432/files#diff-1396a400e5c9f5ccaa44cb20194066eaa919a7e02c558db02005a4a2b67a93b9R2117-R2122). I guess so far this wasn't apparent because all the tests expected a group epoch bump happening. In this case, we didn't want a group epoch bump and hence I could notice the discrepancy.